File: cnl.go

package main
import (
    "fmt"
    "net/http"
    "net/http/httputil"
    "encoding/base64"
    "encoding/hex"
    "regexp"
    "crypto/aes"
    "crypto/cipher"
    "errors"
)

var Debug = 0

func handler(w http.ResponseWriter, r *http.Request){
    if r.Method == "POST" {
        r.ParseForm()
        //debug
        if(Debug == 1){
            requestDump, err := httputil.DumpRequest(r, true)
            if err != nil {
              fmt.Println(err)
            }
            fmt.Println(string(requestDump))
       }
    }else if r.Method == "GET" {
        //fmt.Println("Receive a GET Request")
    }
    var path = r.URL.Path[1:]
    var rep = ""
    if(path == ""){
        rep = flash()
    }else if(path == "jdcheck.js"){
        rep = alive()
    }else if(path == "addcrypted2"){
        //fmt.Println(reflect.TypeOf(r.Form).String())
        addcrypted2(r)
        w.WriteHeader(200)
    }else if(path == "flash/add"){
        add(r)
        w.WriteHeader(200)
    }else{
        fmt.Println("Unsupported path")
        w.WriteHeader(404)
    }
    fmt.Fprintf(w, rep)
}

func flash() string{
    return "JDownloader"
}

func alive() string{
    return "jdownloader=true;\nvar version='10629';\n"
}

func add(r *http.Request){
    // seperator of urls are \n
    method := "add plain"
    urls := r.FormValue("urls")
    source := r.FormValue("source")
    passwords := r.FormValue("passwords")
    fmt.Printf("Methode:%v\n",method)
    fmt.Printf("urls:\n%+v\n", urls)
    fmt.Printf("source:%+v\n",source)
    fmt.Printf("passwords:%+v\n",passwords)
}

func aes_decrypt(key string,data string) (decoded string){
  src :=[]byte(key)
  //encode the key
  dst := make([]byte, hex.DecodedLen(len(src)))
  n, err := hex.Decode(dst, src)
  if err != nil {
    fmt.Println(err)
  }
  //decode the data
  crypted, err := base64.StdEncoding.DecodeString(data)
    if err != nil {
      fmt.Println("decode error:", err)
      return
  }
  cipher_key := dst[:n]
  block, err := aes.NewCipher(cipher_key)
	if err != nil {
		return
  }
  if len(crypted) < aes.BlockSize {
		err = errors.New("Ciphertext block size is too short!")
		return
  }
  mode := cipher.NewCBCDecrypter(block, cipher_key)
  mode.CryptBlocks(crypted, crypted)
  return string(crypted)
}

func addcrypted2(r *http.Request){
    method := "addcryptd2"
    package_name :=r.FormValue("package")
    source := r.FormValue("source")
    passwords := r.FormValue("passwords")
    jk := r.FormValue("jk")
    crypted := r.FormValue("crypted")
    re, _ := regexp.Compile(`\{.?return.?'(\d+)'.?\}`)
    res := re.FindStringSubmatch(jk)
    var key string
    if res != nil{
      key = res[1]
    }else{
      fmt.Println("No Key found")
      return
    }
    urls := aes_decrypt(key,crypted)

    fmt.Printf("Methode:%v\n",method)
    if len(package_name) > 0 {
      fmt.Printf("package:%+v\n",package_name)
    }
    fmt.Printf("source:%+v\n",source)
    fmt.Printf("passwords:%+v\n",passwords)
    fmt.Printf("urls:\n%v\n",urls)
}

func main(){
    fmt.Println("Starting local CNL webserver on localhost:9666")
    http.HandleFunc("/", handler)
    //~ http.HandleFunc("/jdcheck.js", alive)
    //~ http.HandleFunc("/addcrypted2", addcrypted2)
    http.ListenAndServe(":9666",nil)
}

File: cnl.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# pip install pycrypto

import os
import http.server
import socketserver
import re
import logging
from urllib.parse import parse_qs
from Crypto.Cipher import AES
import base64
port = 9666
addr = "127.0.0.1"

logger = logging.getLogger('cnlpy')
logger.setLevel(logging.DEBUG)
cn_log = logging.StreamHandler()
cn_log.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
cn_log.setFormatter(formatter)
logger.addHandler(cn_log)

class ClickHandler(http.server.BaseHTTPRequestHandler):

    def __init__(self, request, client_address, server):
        self.source=''
        self.urls=''
        self.jk=''
        self.password=''
        self.package=''
        self.method=''

        super().__init__(request, client_address, server)

    # not all implemented jet.. kept this funcs for later commits
    def do_GET(self):
        path = self.path.strip("/").lower()
        self.map = [
        (r"add$", self.add),
        #~ (r"addcrypted$", self.addcrypted),
        (r"addcrypted2$", self.addcrypted2),
        #~ (r"flashgot", self.flashgot),
        #~ (r"crossdomain\.xml", self.crossdomain),
        #~ (r"checkSupportForUrl", self.checksupport),
        (r"jdcheck.js", self.jdcheck),
        (r"", self.flash)
        ]
        func = None
        for r, f in self.map:
            if re.match(r"(flash(got)?/?)?"+r, path):
                func = f
                break
        if func:
            try:
                func()
            except Exception as e :
                raise(e)
                self.send_error(500, str(e))
        else:
            self.send_error(404, "Not Found")

    def do_POST(self):
        length=int(self.headers['Content-Length'])
        if not length:
            logger.error("No Content-Lenth in POST Request")
            return

        self.post = parse_qs(self.rfile.read(length).decode('utf-8'))
        self.do_GET()

    def log_message(self, format, *args):
        # silent mode ;-)
        return

    def success(self):
        self.send_response(200, "OK")
        self.end_headers()

    def response(self, response_fn):
        if response_fn is None:
            self.send_error(404, "Not Found")
            return
        try:
            self.send_response(200, "OK")
            self.end_headers()
            self.wfile.write(response_fn.encode())
        except Exception as e:
            self.send_error(500, str(e))
            return

    def debug(self):
        print("DEBUG")
        print("Address: %s"%(self.client_address,))
        print("Path: %s"%(self.path,))
        print("ReqVers: %s"%(self.request_version,))
        print("Headers: %s"%(self.headers,))

    def get_post(self, name, default=""):
        if name in self.post:
            return self.post[name][0]
        else:
            return default

    def aes_decrypt(self,key, data):
        '''
        decrypt cnl2 crypted
        1. base64 decode
        2. base16 decode key
        3. AES 128 (16 block_size) decryption, iv=key
        '''
        key_bytes = 16
        key = base64.b16decode(key)
        assert len(key) == key_bytes
        # Create AES-CBC cipher. Key is IV
        aes = AES.new(key, AES.MODE_CBC, key)
        # Decrypt and return the plaintext.
        return aes.decrypt(base64.b64decode(data)).decode('utf-8').strip()

    def __str__(self):
        rep="Method: %s\n"%(self.method,)
        if self.package:
            rep+="Package: %s\n"%(self.package,)
        rep+="passwords: %s\nsource: %s\nurls:\n%s\n"%(self.passwords,self.source,self.urls)
        return rep

    def add(self):
        self.method="plain"
        self.urls = self.get_post('urls')
        self.passwords = self.get_post('passwords')
        self.source = self.get_post('source')
        print(self)
        self.success()

    def addcrypted2(self):
        self.method="addcrypted2"
        self.passwords = self.get_post("passwords")
        self.source = self.get_post("source")
        self.package = self.get_post("package")
        jk = self.get_post("jk")
        crypted = self.get_post("crypted")
        m=re.search("\{.?return.?'(\d+)'.?\}",jk)
        if m:
            key = m.group(1)
        else:
            print("No Key found...")
            self.send_error(500, str(e))

        self.urls=self.aes_decrypt(key,crypted)
        print(self)
        self.success()

    def flash(self):
        self.response("JDownloader\r\n")

    def jdcheck(self):
        self.response("jdownloader=true;\nvar version='10629';\n\r\n")

if __name__ == "__main__":
    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer((addr, port), ClickHandler) as httpd:
        print("serving at port", port)
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass
        finally:
            httpd.server_close()
